numpy/ma/core.py

```diff
@@ -2978,9 +2976,19 @@ def __array_finalize__(self, obj):
             # heuristic it's not bad.) In all other cases, we make a copy of
             # the mask, so that future modifications to 'self' do not end up
             # side-effecting 'obj' as well.
-            if (obj.__array_interface__["data"][0]
+            if (_mask is not nomask and obj.__array_interface__["data"][0]
                     != self.__array_interface__["data"][0]):
-                _mask = _mask.copy()
+                # We should make a copy. But we could get here via astype,
+                # in which case the mask might need a new dtype as well
+                # (e.g., changing to or from a structured dtype), and the
+                # order could have changed. Cover these with astype if the
+                # dtype is not the same (as it would be in slices).
+                order = "F" if self.flags.f_contiguous else "C"
```

Comment:
This changes the meaning of `order='K'` compared to my PR

Reply:
Yes, I cannot recover what was given. I don't know that it matters, though, that we preserve possibly different order for the data and the mask...

Reply:
I suppose I'm questioning whether `__array_finalize__` is the right place for this - that's where we have the least information, and it's perhaps possible that we got there through a different mechanism than `astype`.

Reply:
This is specifically meant for subclasses to update information after a new view or different type cast is made, which is why there is the same `_update` call, etc. Personally, I don't see a real downside, and any simplification helps in getting closer to a new `MaskedArray` that uses `__array_ufunc__` (which means many of the numerical method overrides can be removed).

Reply:
How about:
```
if self.flags.c_contiguous
    order = "C"
elif self.flags.f_contiguous:
    order = "F"
else:
    order='K'
```